src/lib.rs

```diff
@@ -1644,4 +1652,463 @@ mod tests {
     test_ivc_base_with::<bn256::Point, grumpkin::Point>();
     test_ivc_base_with::<secp256k1::Point, secq256k1::Point>();
   }
+
+  fn print_constraints_name_on_error_index<G1, G2, C1>(err: &NovaError, c_primary: &C1)
+  where
+    G1: Group<Base = <G2 as Group>::Scalar>,
+    G2: Group<Base = <G1 as Group>::Scalar>,
+    C1: StepCircuit<G1::Scalar>,
+  {
+    match err {
+      NovaError::UnSatIndex(index) => {
+        let augmented_circuit_params_primary =
+          NovaAugmentedCircuitParams::new(BN_LIMB_WIDTH, BN_N_LIMBS, true);
+
+        // let (mut circuit_primary, z0_primary) = HeapifyCircuit::new(ro_consts);
+        let ro_consts_circuit_primary: ROConstantsCircuit<G2> = ROConstantsCircuit::<G2>::default();
+        let circuit_primary: NovaAugmentedCircuit<'_, G2, C1> = NovaAugmentedCircuit::new(
+          &augmented_circuit_params_primary,
+          None,
+          c_primary,
+          ro_consts_circuit_primary,
+        );
+        // let mut cs: ShapeCS<G1> = ShapeCS::new();
+        // let _ = circuit_primary.synthesize(&mut cs);
+        let mut cs: TestShapeCS<G1> = TestShapeCS::new();
+        let _ = circuit_primary.synthesize(&mut cs);
+        cs.constraints
+          .get(*index)
+          .tap_some(|constraint| println!("failed at constraint {}", constraint.3));
+      }
+      error => unimplemented!("{:?}", error),
+    }
+  }
+
+  #[test]
+  fn test_ivc_rwlookup() {
+    type G1 = pasta_curves::pallas::Point;
+    type G2 = pasta_curves::vesta::Point;
+
+    // rw lookup to serve as a non-deterministic advices.
+    #[derive(Clone)]
+    struct HeapifyCircuit<G1: Group, G2: Group>
+    where
+      <G1 as Group>::Scalar: Ord,
+      G1: Group<Base = <G2 as Group>::Scalar>,
+      G2: Group<Base = <G1 as Group>::Scalar>,
+    {
+      lookup_trace: LookupTrace<G1>,
+      ro_consts: ROConstantsCircuit<G2>,
+      max_value_bits: usize,
+      _phantom: PhantomData<G2>,
+    }
+
+    impl<G1: Group, G2: Group> HeapifyCircuit<G1, G2>
+    where
+      <G1 as Group>::Scalar: Ord,
+      G1: Group<Base = <G2 as Group>::Scalar>,
+      G2: Group<Base = <G1 as Group>::Scalar>,
+    {
+      fn new(
+        initial_table: &Lookup<G1::Scalar>,
+        ro_consts_circuit: ROConstantsCircuit<G2>,
+      ) -> (Vec<Self>, Lookup<G1::Scalar>, G1::Scalar) {
+        let n = initial_table.table_size();
+        let initial_table = initial_table.clone();
```

Comment:
You should instead take the initial table by value ([`C-CALLER-CONTROL`](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control))

Reply:
Fixed